### ISSUE TITLE
response reading failure is accompanied by 2xx reply code

### ISSUE THREAD

**Issue by vr009** [curl/curl#8356]
### I did this

I have the next scenario. I use libcurl for SMTP client app. Normally if I get some SMTP error, I need to catch and handle the CURL code `CURLE_RECV_ERROR` or `CURLE_SEND_ERROR`. Generally that works well as expected, but I watched the next situation. If mail server breaks the connection after receiving `RCPT TO:` or `MAIL FROM:` command, the curl code is `CURLE_RECV_ERROR`, at the same time the reply code is 250 and there is a message `response reading failed` collected in error buffer.

### I expected the following

It looks as a problem in libcurl: there is no reliable way to obtain an SMTP reply code. It is especially strange to get 2xx reply code for failure at reading.

### curl/libcurl version

curl 7.81.0 (x86_64-apple-darwin20.6.0) libcurl/7.81.0 

### operating system

Darwin 20.6.0 Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:42 PDT 2021; root:xnu-7195.141.8~1/RELEASE_X86_64 x86_64

### Reproducing the problem
Client code is like this:

```c
#include <stdio.h>
#include <string.h>
#include <curl/curl.h>
#include <stdlib.h>

#define FROM_ADDR    "<sender@example.org>"
#define TO_ADDR      "breakconnect@example.org"

int main(void)
{
	CURL *curl;
	CURLcode res = CURLE_OK;
	int status;
	char *error_buf = calloc(CURL_ERROR_SIZE, sizeof(char));
	struct curl_slist *recipients = NULL;

	curl = curl_easy_init();
	if(curl) {
		curl_easy_setopt(curl, CURLOPT_URL, "smtp://<ip>:<port>");
		curl_easy_setopt(curl, CURLOPT_MAIL_FROM, FROM_ADDR);
		recipients = curl_slist_append(recipients, TO_ADDR);
		curl_easy_setopt(curl, CURLOPT_MAIL_RCPT, recipients);
		curl_easy_setopt(curl, CURLOPT_UPLOAD, 1L);
		curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, error_buf);

		/* Send the message */
		res = curl_easy_perform(curl);

		long longval = 0;
		switch (res) {
			case CURLE_OK:
				curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &longval);
				status = (int) longval;
				fprintf(stdout, "curl_easy_perform() succeed: %d\n",
				        status);
				/*
				 * For sure that the status code here doesn't affect
				 * the status code below.
				 */
				status = 0;
				break;
			case CURLE_SEND_ERROR:
			case CURLE_RECV_ERROR:
				curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &longval);
				status = (int) longval;
				/*
				 * When client can not read bytes from another side that
				 * leads to a message about failed response reading but
				 * response code stays 250, that seems to be not correct.
				 */
				fprintf(stdout, "curl_easy_perform() failed. \nCode status: %d\nDescription: %s\n",
				        status, error_buf);
				break;
			default: {
				free(error_buf);
				return -1;
			}
		}
		curl_slist_free_all(recipients);
		curl_easy_cleanup(curl);
	}
	free(error_buf);
	return (int)res;
}

```
Lua written mail server:

```lua
local socket = require('socket')

local function smtp_h(s)
    s:write('220 localhost ESMTP Tarantool\r\n')
    local l
    local mail = {rcpt = {}}
    while true do
        l = s:read('\r\n')
        if l:find('EHLO') then
            s:write('250-localhost Hello localhost.lan [127.0.0.1]\r\n')
            s:write('250-SIZE 52428800\r\n')
            s:write('250-8BITMIME\r\n')
            s:write('250-PIPELINING\r\n')
            s:write('250-CHUNKING\r\n')
            s:write('250-PRDR\r\n')
            s:write('250 HELP\r\n')
        elseif l:find('MAIL FROM:') then
            mail.from = l:sub(11):sub(1, -3)
            if l:find('breakconnect') then
                return
            else
                s:write('250 OK\r\n')
            end
        elseif l:find('RCPT TO:') then
            mail.rcpt[#mail.rcpt + 1] = l:sub(9):sub(1, -3)
            if l:find('breakconnect') then
                return
            else
                s:write('250 OK\r\n')
            end
        elseif l == 'DATA\r\n' then
            s:write('354 Enter message, ending with "." on a line by itself\r\n')
            while true do
                local l = s:read('\r\n')
                if l == '.\r\n' then
                    break
                end
                mail.text = (mail.text or '') .. l
            end
            mail = {rcpt = {}}
            s:write('250 OK\r\n')
        elseif l:find('QUIT') then
            return
        elseif l ~= nil then
            s:write('502 Not implemented')
        else
            return
        end
    end
end

local server = socket.tcp_server('127.0.0.1', 0, smtp_h)
local addr = 'smtp://127.0.0.1:' .. server:name().port

```

The original problem and discussion on this can be found [here](https://github.com/tarantool/smtp/pull/53).

**Comment by bagder** [comment]
In the man page for `CURLINFO_RESPONSE_CODE` it is documented to work like this:

> Pass a pointer to a long to receive the last received [...] response code.

So isn't 250 the last received response code?

**Comment by Totktonada** [comment]
I'm looking on it this way: after `curl_easy_perform()` I should be able to differentiate the following situations:

1. A network failure.
2. An unexpected server response (it seems, it is what we have in the reproducer above).
3. A valid server response with particular SMTP reply code.

While I'm not so bother about distinguishing between 1 and 2, the latter two looks important (a caller may want to handle those cases differently). Now I don't see a way to differentiate 2 and 3 aside of `strcmp()` on the libcurl's error message.

If `CURLINFO_RESPONSE_CODE` is not suitable to obtain an SMTP reply code **only when it has a sense** (and some indicator otherwise), there should be another way to get this information, I think.

**Comment by bagder** [comment]
(Sorry but I don't know how to run that test server code and I'm not quite following what it shows.)

Doesn't case 2 and 3 differ in the way so case 2 returns and error and case 3 returns `CURLE_OK` ?

**Comment by jzakrzewski** [comment]
I may be mistaken but I don't think server responding with SMTP Error code would cause `curl_easy_perform()` to report an error. That means the network transfer is completed but you need to check the response code to know if server accepted it.

On the other hand manual says:
```
CURLE_SEND_ERROR (55)

Failed sending network data.

CURLE_RECV_ERROR (56)

Failure with receiving network data.
```

Which to me mean that those two indicate some kind of network issues (including unexpected server disconnect, like you do).
Am I wrong?

**Comment by bagder** [comment]
The SMTP code returns `CURLE_RECV_ERROR` in two places where it clearly is not a network problem, which looks like a problem. Is that the problem we're talking about?

**Comment by Totktonada** [comment]
> (Sorry but I don't know how to run that test server code and I'm not quite following what it shows.)

Hm, it seems, the server code is specific for the 'socket' module offered by tarantool. I would ask @vr009 to provide a server code on C or at least using 'luasocket' module to reproduce with any Lua interpreter.

> Doesn't case 2 and 3 differ in the way so case 2 returns error and case 3 returns `CURLE_OK` ?

In both cases we can receive `CURLE_RECV_ERROR`. I guess it is better to point our [error handling code](https://github.com/tarantool/smtp/blob/12c4142e14c32a7e3c578d98b651f6ffbefa01d6/smtp/smtpc.c#L453-L468) — it should describe the problem better than my words.

> The SMTP code returns `CURLE_RECV_ERROR` in two places where it clearly is not a network problem, which looks like a problem. Is that the problem we're talking about?

It looks quite relevant at least. If we would have a constant for `curl_easy_perform()` return value, which states that a non-2xx SMTP reply returned by a server (and nothing more critical occurs), it would solve the problem, I guess.

----

I found the situation quite similar to #4310 in the sense that we receive a general `CURLE_RECV_ERROR` for different situations, but there should be a code that say us 'you can use `CURLINFO_RESPONSE_CODE` to obtain a valid SMTP reply code'.

**Comment by vr009** [comment]
I'm sorry for the server code above. If you don't mind I leave the python implementation of the test server that also reproduces this situation.

```python
import socketserver

class Mailserver(socketserver.StreamRequestHandler):
    def handle(self):
        self.wfile.write(bytes('220 localhost ESMTP\r\n', 'utf-8'))
        while True:
            l = str(self.rfile.readline().strip())
            cmd = l[2:len(l) - 1]
            if cmd.find('EHLO') != -1:
                self.wfile.write(bytes('220 localhost ESMTP Tarantool\r\n', 'utf-8'))
                self.wfile.write(bytes('250-localhost Hello localhost.lan [127.0.0.1]\r\n', 'utf-8'))
                self.wfile.write(bytes('250-SIZE 52428800\r\n', 'utf-8'))
                self.wfile.write(bytes('250-8BITMIME\r\n', 'utf-8'))
                self.wfile.write(bytes('250-PIPELINING\r\n', 'utf-8'))
                self.wfile.write(bytes('250-CHUNKING\r\n', 'utf-8'))
                self.wfile.write(bytes('250-PRDR\r\n', 'utf-8'))
                self.wfile.write(bytes('250 HELP\r\n', 'utf-8'))
            elif cmd.find('MAIL FROM:') != -1:
                mail_from = cmd[11:len(cmd) - 1]
                if mail_from.find('breakconnect') != -1:
                    return
            elif cmd.find('RCPT TO:') != -1:
                rcpt = cmd[9:len(l) - 1]
                if rcpt.find('breakconnect') != -1:
                    return
            elif cmd == 'DATA\r\n':
                return
            elif cmd.find('QUIT') != -1:
                return
            else:
                return


HOST, PORT = "127.0.0.1", 9999
with socketserver.TCPServer((HOST, PORT), Mailserver) as server:
    server.serve_forever()
```

**Comment by Totktonada** [comment]
Sorry for requesting more attention to this issue: I want to understand its status.

* Whether it is confirmed by the author?
* Whether more details are necessary from us?

**Comment by bagder** [comment]
I'm still struggling to understand what the problem is.

**Comment by Totktonada** [comment]
Sorry if my description is not well written. I'll try to be concise and concrete.

I want to distinguish errors from `curl_easy_perform()`, which serves an SMTP request, and pass them to a caller. A caller should be able to programmatically decide, whether to retry the request or not. So at least we should differentiate:

1. Availability (network, server) errors.
2. SMPT server response with particular reply code.
3. All other (usually not expected) situations: say, incorrect server response.

They are not always programmatically distinguishable (without `strcmp()` hack under `CURLE_RECV_ERROR` case, which is unlikely durable).

In other words, if I got `CURLE_RECV_ERROR` — it should be an error from `read()` or something of this kind.

And there should be a `curl_easy_perform()` return value (say, `CURLE_SMTP_UNSUCCESSFUL_REPLY`), which guarantees that `curl_easy_getinfo(req->easy, CURLINFO_RESPONSE_CODE, &reply_code);` will return an actual reply code from the server. Alternative: libcurl should have documented library-specific reply codes, which are set by the library itself (as it does for the timeout error in context of HTTP).

The main point: ability to programmatically distinguish different failure situations.

**Comment by bagder** [comment]
> And there should be a curl_easy_perform() return value (say, CURLE_SMTP_UNSUCCESSFUL_REPLY), which guarantees that curl_easy_getinfo(req->easy, CURLINFO_RESPONSE_CODE, &reply_code); will return an actual reply code from the server

We can't do that. `CURLINFO_RESPONSE_CODE` is already documented, implemented and relied on to work the way it currently works. It is the last response code from the server that curl saw.

What we *can* do however, is to make sure that `CURL_RECV_ERROR` is not used for cases where curl just think the server reply isn't satisfactory and instead rather use `CURLE_WEIRD_SERVER_REPLY` for those errors. Then you could treat `CURL_RECV_ERROR` as an error that means there was a problem with getting data from the connection.

**Comment by bagder** [comment]
#8506 is an attempt to do exactly that and I would value your feedback on that approach!

**Comment by Totktonada** [comment]
Thanks!

I have one more question, though. How should I differentiate a 'real' read error (a connectivity problem) from non-2xx SMTP reply code?

**Comment by bagder** [comment]
A "real" read error would return `CURL_RECV_ERROR`.

**Comment by Totktonada** [comment]
Okay. So whether `curl_easy_getinfo(req->easy, CURLINFO_RESPONSE_CODE, &reply_code);` always return a 'real' server's reply code, when `CURLE_WEIRD_SERVER_REPLY` is returned from `curl_easy_perform()`?

Or, if a server returns some garbage (1000 as a reply code, no reply code at all, syntactically wrong response), we'll also get `CURLE_WEIRD_SERVER_REPLY`? If so, how to differentiate this situation from the "real server's reply code" one?

(Sorry, it is hard to me to deduce it from the patch — I have no good knowledge around libcurl's codebase.)

**Comment by bagder** [comment]
> Or, if a server returns some garbage (1000 as a reply code, no reply code at all, syntactically wrong response), we'll also get CURLE_WEIRD_SERVER_REPLY?

Yes.

>  If so, how to differentiate this situation from the "real server's reply code" one?

How is 1000 not "the real code" then if the server returned that?

**Comment by Totktonada** [comment]
Okay, 1000 is the real server's reply code (just incorrect). `..._getinfo()` will return it if server respond us so.

I can write a table of situations:

| Situation | `curl_easy_perform()` retval | reply code from `curl_easy_getinfo()` |
| ---- | ---- | ---- |
| `read()` error | CURL_RECV_ERROR | undefined |
| 2xx reply code | CURLE_OK | 2xx, from server response |
| non-2xx reply code | CURLE_WEIRD_SERVER_REPLY | non-2xx, from server response |
| reply code from outside of the range defined by the standard | CURLE_WEIRD_SERVER_REPLY | from server response |
| malformed response | CURLE_WEIRD_SERVER_REPLY | ? |
| malformed response, but the problem is after parsing the reply code | CURLE_WEIRD_SERVER_REPLY | ? |

What should I expect in the last two lines and whether other lines are correct (according to the implementation from PR #8506)?

**Comment by bagder** [comment]
1. If the transfer performs fine, it returns `CURLE_OK`
2. For read() errors it returns `CURL_RECV_ERROR`
3. every response code returned is stored (overwriting the previous one), and is retrievable with `curl_easy_getinfo()`
4. if libcurl can't continue due to a non-recognized response code or other *protocol peculiarities* it returns `CURLE_WEIRD_SERVER_REPLY`

**Comment by Totktonada** [comment]
So line 3 (non-2xx reply) from the table above is `CURLE_WEIRD_SERVER_REPLY` (and we can acquire a reply code) and non-recognized response is `CURLE_WEIRD_SERVER_REPLY` (and we should not attempt to acquire a reply code). And there is not way to distinguish one `CURLE_WEIRD_SERVER_REPLY` from another `CURLE_WEIRD_SERVER_REPLY`. Correct?

**Comment by bagder** [comment]
> there is not way to distinguish one CURLE_WEIRD_SERVER_REPLY from another CURLE_WEIRD_SERVER_REPLY. Correct?

The error buffer contents will be different and you can use the debugfunction to figure out in more detail exactly what happened, but the return codes will be the same yes.

**Comment by Totktonada** [comment]
So trying to avoid the [`strcmp()` hack](https://github.com/tarantool/smtp/blob/12c4142e14c32a7e3c578d98b651f6ffbefa01d6/smtp/smtpc.c#L453-L468) we'll end up with the similar `strcmp()` hack.

Okay, if it is considered as a proper error handling way, is there a list of strings I should use to differentiate situations, when I should and shout not extract a reply code?

**Comment by bagder** [comment]
What are you going to strcmp?

**Comment by Totktonada** [comment]
The error buffer (I hope it'll be enough).

**Comment by bagder** [comment]
I don't understand what kind of situations you try to differentiate with this when libcurl doesn't understand the response and returns this specific error. You then want to try to interpret or "outsmart" what libcurl returns and draw further conclusions about the error? What are you going to do with those assumptions? Why does it matter to you what exact detail it was that libcurl couldn't deal with in the protocol communication?

libcurl will not guarantee specific error buffer contents. Relying on that will be very fragile and likely break in future releases.

**Comment by Totktonada** [comment]
I have a field with the reply code in the error, which is reported to a user. I can't just fill it with what libcurl reports, because it returns 250 for a definitely malformed response. I should place -1 or some other specical value here in the case.

Just checked and if I'm doing everything correct, if I close the connection on the server right before writting a reply code, libcurl returns `CURLE_RECV_ERROR` (as before the patch as welll as after applying it). Attempt to get a reply code from libcurl gave 250 in the case.

So I doubt what condition should I use to either fill my reply code with what libcurl returns or fill it with -1.

Now the condition is the following:

* `rc == CURLE_OK` → use the reply code from libcurl
* `(rc == CURLE_SEND_ERROR || rc == CURLE_RECV_ERROR) && strcmp(error_buf, "response reading failed") != 0` → use the reply code from libcurl
* `(rc == CURLE_SEND_ERROR || rc == CURLE_RECV_ERROR) && strcmp(<...>) == 0` → set -1
* Otherwise: set -1.

As you confirmed, relying on the error buffer content is the fragile solution (and possibly incomplete). It is why this issue appears. Is there a better way to make the programmatic decision, whether the reply code from libcurl can be used or should be considered as undefined?

Or, better, can libcurl set the reply code to -1 (or other special value), when acquiring a reply code has no sense?

**Comment by jay** [comment]
libcurl does not guarantee that a protocol-specific [response code](https://curl.se/libcurl/c/CURLINFO_RESPONSE_CODE.html) will be set.

>_The stored value will be zero if no server response code has been received._

I understand you may want more distinct error codes but I think #8506 is enough. If libcurl returns an error --no matter the error-- check the error buffer first and if that's empty then fall back on [curl_easy_strerror](https://curl.se/libcurl/c/curl_easy_strerror.html).

**Comment by Totktonada** [comment]
> if I close the connection on the server right before writting a reply code

Sorry, here I made a mistake. I closed the connection (on a server) after answering 250 to `EHLO` (right after reading `MAIL FROM` from a client). So last reply is actually 250. It is just not, hm, a 'final' reply code. It misleads me initially.

So my condition could be the following:

* `rc == CURLE_OK` → use the reply code from libcurl
* `(rc == CURLE_SEND_ERROR || rc == CURLE_RECV_ERROR) && reply_code in (250, 251, 252)` → use the reply code from libcurl
* `(rc == CURLE_SEND_ERROR || rc == CURLE_RECV_ERROR) && reply_code not in (250, 251, 252)` → set -1.
* Otherwise: set -1.

Is it correct way to determine, whether the *whole* attempt to send a email is failed with a particular reply code? Maybe there is a better way?

**Comment by Totktonada** [comment]
(I understood that this discussion becomes a bit tiresome. I appreciate friendliness and patience of participants and want to say 'Thank you!' to all who engaged here.)

It seems, we stuck here again. Nobody likes such freezed discussions and I think we all would like to move forward somehow. I would like to propose a solution: add more details into the `CURLINFO_RESPONSE_CODE` description.

I think that Daniel has own vision how errors should be handled in an application's code and an example of 'the only right way' to handle errors from the author should solve all or most of questions.

I would like to highlight things that were confusing for me:

* Whether the response code value is defined after any `curl_easy_perform()` call or only after particular set of `curl_easy_perform()` return codes?
* Sending a email using the SMTP protocol assumes several requests from the client and the same amount of responses from the server. The response code returned by the library is not necessarily from a response to the last client's request. It is *last received* response code, while the client not necessarily received the response to the latest request and, even if it is received, it can be malformed.

Can we go this way?

**Comment by bagder** [comment]
You say you want to pass the response code to the user. I'm not sure what user that is, but what is this user doing with a response code to a command that it doesn't know was sent? What if a future version of libcurl changes the order of commands or adds a command to the sequence?

> Is it correct way to determine, whether the whole attempt to send a email is failed with a particular reply code? Maybe there is a better way?

libcurl doesn't provide that level of granularity in its return codes, so you cannot tell if the failure to send was exactly because of a particular reply code or because of something else that libcurl couldn't handle. To be able to do what you ask 100% reliably, you would need libcurl to have a unique error code for "failed to use the reply code" and there isn't one now. All other ways will be doing more or less assumptions that may or may not break as libcurl internals change over time.

> Whether the response code value is defined after any curl_easy_perform() call or only after particular set of curl_easy_perform() return codes?

The documentation for [CURLINFO_RESPONSE_CODE](https://curl.se/libcurl/c/CURLINFO_RESPONSE_CODE.html) explains that the value you get is "the last received ... SMTP ... response code". That's totally independent of what was returned as a result of the transfer, and in fact the wording in the documentation is the promise we make for applications on how this works, has worked like and what we need to stick to going forward.

## forward

I personally cannot condone any approach that checks or relies on the contents of error messages. That's going to break at some point. I mean, everyone is free to do what they want, but that's a fragile way.

I think the #8506 approach is pretty close to what you ask for, even if it doesn't go all the way. With that, you can ignore the response code for all return codes except for `CURLE_WEIRD_SERVER_REPLY`, which has only the small glitch that it will also return this error for a few other parsing problems. I suppose that takes us back to the original question: what's the user supposed to do with the response code when libcurl has failed to send mail?

**Comment by Totktonada** [comment]
Sorry for the late answer and thanks for the patience!

----

> what's the user supposed to do with the response code when libcurl has failed to send mail?

I want to give a user a finite amount of well defined request outcomes and let her/him to decide whether to retry, whether to save the request in a repair queue and notify the operation team, whether to just log the failure and discard the request.

Example: in one of applications I have two queues of emails to send with different retry policy. The first queue attempt to deliver a email over any failure except situations, when we know that a retry will not help (and it is better to fail faster). This queue is used, for example, for emails sent as part of a registration process. Another queue is for OTP codes: we don't retry them, because the application's user able to ask to re-send the code if something goes wrong.

In my mind, the ideal picture is the following: I'll describe the next situations in our smtp module docs.

* Success with a reply code (250, 251, 252).
* A network failure (likely retryable if a caller don't bother about a little probability of a duplicate request).
* A bad response from a server (likely something for a repair queue / operation team).
* Failure with a reply code (here I would just link the RFC 5321 document).
* Internal errors like OOM.
* Arguments validatoin error and so on (programmer errors).

----

> you can ignore the response code for all return codes except for `CURLE_WEIRD_SERVER_REPLY`, which has only the small glitch that it will also return this error for a few other parsing problems.

But why? Technically I have no return code, which allows me to lean on the reply code.

Do you think I'm too pedantic there? But how I'll describe it in our documentation: "a reply code may or may not represent actual source of the request failure"? I read it as "a reply code is useless for automatic decisions".

**Comment by bagder** [comment]
For me you can be as pedantic as you want, that's not my concern. My main concern here is that we need to stick to API promises we have already made since a long time. If there had been any doubts or if the behavior was unclear or flaky in some way then clarifying would've been good. In this case however, it seems like libcurl behaves like the documentation states. We can't change that without risking we break existing applications.

I'm at the end of the road here. I don't know what you ask from libcurl.

Your issue seems to be that the (documented) behavior is unhelpful. Which is a view I can understand after your explanations, but I think the only reasonable way out of this situation is to *add* something to libcurl to helps you get the data the way you want it.

**Comment by Totktonada** [comment]
I guess the API was initially designed for HTTP reply codes, where we usually have only one request-response pair. Except redirections. That just comes into the mind: there is likely the same problem with HTTP redirections.

Of course, the only way to change existing public API is to extend it.

It seems, I should accept my fail. You said that you still don't understand what I asked. So unlikely I can ask you to implement a thing that is vague for you.

Thank you for your time anyway!

**Comment by bagder** [comment]
I think I understand it a little bit better now, but I would still need the exact requested functionality to be spelled out with detail in order to even consider working on implementing something. It might even be even better if someone like you, who actually need/want this, would implement it!

**Comment by Totktonada** [comment]
You want it to be rephrased as a request for a new functionality (plus API proposal) rather than description of a problem in the existing one, correct?

In my mind it is something like the following.

> CURLINFO_FINAL_RESPONSE_CODE - get the final response code
>
> ```c
> CURLcode curl_easy_getinfo(CURL *handle, CURLINFO_FINAL_RESPONSE_CODE, long *codep);
> ```
>
> Pass a pointer to a long to receive the final SMTP reply code or zero if the last reply is not received / can't be parsed. To illustrate the difference with `CURLINFO_RESPONSE_CODE`, consider the following client-server iteraction:
>
> ```
> S: [a client connected]
> S: 220 foo.com Simple Mail Transfer Service Ready
> C: EHLO bar.com
> S: 250-foo.com greets bar.com
> S: 250-8BITMIME
> S: 250-SIZE
> S: 250-DSN
> S: 250 HELP
> C: MAIL FROM:<Smith@bar.com>
> S: [unexpectedly closed the client's connection]
> ```
>
> (The scenarion is based on RFC 5321 [appendix D.1](https://datatracker.ietf.org/doc/html/rfc5321#appendix-D.1).)
>
> `CURLINFO_RESPONSE_CODE` is 250, because the last server reply is 250. However the client sends a request and receives EOL instead of a response. So `CURLINFO_FINAL_RESPONSE_CODE` is zero.
>
> The idea of the final response code is to never receive non-zero result in a case, when something aside a last response code leads to a request failure. In other words, the final code is either the reason of the fail or zero.

However it is not the only variant. Say, a list of all reply codes would also do the job, if a new zero entry will be added before each request to the server. But it would be less convenient for error handling on caller's side, I guess.

----

I would like to try to implement a PoC, but can't promise that I'll do it in a near future.

**Comment by bagder** [comment]
I'm closing this issue based on the simple fact that curl is actually doing what it is documented to here as I believe this long discussion has proven. The conclusion is rather that libcurl should probably offer something different/better/more to satisfy the desires and needs of SMTP applications, but that's not a bug.

**Comment by Totktonada** [comment]
I agree, this is the feature request.
